python-package/lightgbm/basic.py

```diff
@@ -502,14 +502,19 @@ def c_int_array(data):
 
 
 def _get_bad_pandas_dtypes(dtypes):
-    pandas_dtype_mapper = {'int8': 'int', 'int16': 'int', 'int32': 'int',
-                           'int64': 'int', 'uint8': 'int', 'uint16': 'int',
-                           'uint32': 'int', 'uint64': 'int', 'bool': 'int',
-                           'float16': 'float', 'float32': 'float', 'float64': 'float'}
-    bad_indices = [i for i, dtype in enumerate(dtypes) if (dtype.name not in pandas_dtype_mapper
-                                                           and (not is_dtype_sparse(dtype)
-                                                                or dtype.subtype.name not in pandas_dtype_mapper))]
-    return bad_indices
+    return [i for i, dtype in enumerate(dtypes) if not is_numeric_dtype(dtype)]
```

Comment:
By switching from fixed mapper to general `is_numeric_dtype()` function we are allowing to pass new types that were not allowed previously. For example, `np.complex64` and `np.float128`. In other words, every subclass of `numpy.number`:
```
>>> subdtypes(np.generic)

[numpy.generic,
 [[numpy.number,
   [[numpy.integer,
     [[numpy.signedinteger,
       [numpy.int8,
        numpy.int16,
        numpy.int32,
        numpy.int64,
        numpy.longlong,
        numpy.timedelta64]],
      [numpy.unsignedinteger,
       [numpy.uint8,
        numpy.uint16,
        numpy.uint32,
        numpy.uint64,
        numpy.ulonglong]]]],
    [numpy.inexact,
     [[numpy.floating,
       [numpy.float16, numpy.float32, numpy.float64, numpy.float128]],
      [numpy.complexfloating,
       [numpy.complex64, numpy.complex128, numpy.complex256]]]]]],
  [numpy.flexible,
   [[numpy.character, [numpy.bytes_, numpy.str_]],
    [numpy.void, [numpy.record]]]],
  numpy.bool_,
  numpy.datetime64,
  numpy.object_]]
```
https://pandas.pydata.org/docs/user_guide/basics.html#:~:text=All%20NumPy%20dtypes%20are%20subclasses%20of%20numpy.generic%3A

Reply:
Oh you're right, I hadn't thought about that. Do you think it'd be better to explicitly list all the allowed dtypes and just do something like: `[i for i, dtype in enumerate(dtypes) if isinstance(dtype, allowed_dtypes)`

Reply:
I used the idea of what [is_numeric_dtype](https://github.com/pandas-dev/pandas/blob/66e3805b8cabe977f40c05259cc3fcf7ead5687d/pandas/core/dtypes/common.py#L1188) does and checked for more specific subclasses (np.integer, np.floating and np.bool_) in 98325e920def79a565b36fdcf7ede8a2b70c381e. Running `{dtype: subdtypes(dtype) for dtype in (np.integer, np.floating, np.bool_)}` with the subdtypes function in the pandas link you posted returns:
```
{numpy.integer: [numpy.integer,
  [[numpy.signedinteger,
    [numpy.int8,
     numpy.int16,
     numpy.int32,
     numpy.int64,
     numpy.longlong,
     numpy.timedelta64]],
   [numpy.unsignedinteger,
    [numpy.uint8,
     numpy.uint16,
     numpy.uint32,
     numpy.uint64,
     numpy.ulonglong]]]],
 numpy.floating: [numpy.floating,
  [numpy.float16, numpy.float32, numpy.float64, numpy.float128]],
 numpy.bool_: numpy.bool_}
```
and I removed the np.float128 and np.timedelta64. Let me know what you think